doc/design/backward.md

```diff
@@ -0,0 +1,156 @@
+# Backward Building
+
+## Motivation
+
+In Neural Network, most models are solved by the backpropagation algorithm(known as **BP**) at present. Technically, BP calculates the gradient of the loss function, then propagates it back through the networks following the chain rule. However, when configuring the model structure, users do not need to define the backward part. So a mechanism is required by the framework which can complete the model's backward part automatically according to the given forward part.
+
+When implementing a specific `op`, the developer is also asked to implement its backward version, called `grad_op`. A `grad_op` takes gradients of its corresponding `op`'s outputs, and calculate gradients of the `op`'s inputs. During the building of a model's backward part, the framework creates each forward `op`'s `grad_op`, and then string them together in reverse order of forwarding part. In this way, gradients spread from the end to the beginning of the model, in another word, from the loss to parameters.
```

Comment:
`grad_op` is not the backward version of the forward operator.

We maintain a mapping between an operator and operators that will produce its gradient. It is not a `one-to-one` mapping. Operators in `backward stage`(I do not think there should be a `backward stage`.However, to make it easy to understand, I just assume operators in Block can be split into two stages) can be used in forward stage.

Reply:
Great idea! I think we can merge this pr first and I will create an issue to record your idea. I'm going to refine it later.